### DISCUSSION TITLE
Should GitHub be the primary code source and contribution channel?

### DISCUSSION THREAD

**Discussion by verdverm** [cue-lang/cue#1197] [General]
Was reading this blog post (https://matt-rickard.com/a-practical-guide-to-growing-your-open-source-project-part-1-contributors/)

and thought about

- how Gerrithub sucks (imho), clicking around removes any Cuelang context and I have to then add filters at page changes, there are more frustrations...
- Gerrit workflow is unfamiliar to most and requires non standard tools
- these create higher barriers to contribution

So should we drop Gerrit and work from GitHub?

**Comment by myitcv** [comment]
@verdverm as you are aware people can still contribute to the project via PRs. We should make that clearer on the [contribution page](https://github.com/cue-lang/cue/blob/master/doc/contribute.md). 

I can't see a scenario in which we would stop supporting PR submissions. Equally I can't readily see a scenario in which we would stop using Gerrit either because it offers so much for more complex changes. Each mechanism has different benefits/costs, that's why we support both.

If you have any constructive suggestions on areas of GerritHub that need fixing/could be improved, we can pass those on.

**Reply by myitcv** [reply]
https://github.com/cue-lang/cue/issues/1198

**Reply by verdverm** [reply]
There is a lot to be desired in the UX/DX of Gerrit UIs. The google hosted one of prior was better than GerritHub, but both are far behind GitHub, GitLab, and Bitbucket.

Two way seems mostly fine if people are ok with the additional complexity. I think the chief complaint I have about the current setup is that work (code changes) that happens via gerrit are not found on GitHub. Do changes on Gerrit trigger the GitHub workflows?

**Reply by myitcv** [reply]
> There is a lot to be desired in the UX/DX of Gerrit UIs.

If you prefer to use GitHub, that channel will always remain open. Other people prefer Gerrit. That's fine - people can choose!

> Two way seems mostly fine if people are ok with the additional complexity

But the contributor is not exposed to any additional complexity, right? It's complexity for us as a project (and other projects like it, for example the Go project), but that's fine to my mind.

> I think the chief complaint I have about the current setup is that work (code changes) that happens via gerrit are not found on GitHub

I don't understand this point. By all means please contribute via GitHub if that's what you prefer!

> Do changes on Gerrit trigger the GitHub workflows?

The "trybots" can be triggered by project maintainers, similarly with pull requests.

**Reply by verdverm** [reply]
> > I think the chief complaint I have about the current setup is that work (code changes) that happens via gerrit are not found on GitHub

> I don't understand this point. By all means please contribute via GitHub if that's what you prefer!

For any code changes being pushed to Gerrit, one cannot discover, follow, or review them from GitHub.

Keeping up with Gerrit is more difficult after the move because the UX got worse.

**Reply by myitcv** [reply]
> For any code changes being pushed to Gerrit, one cannot discover, follow, or review them from GitHub.

Neither system is perfect, nor is the combination of the two a perfect setup; the two approaches satisfy different requirements.

> Keeping up with Gerrit is more difficult after the move because the UX got worse.

Please can you provide concrete examples of what you find frustrating about the UX?

**Reply by verdverm** [reply]
> Please can you provide concrete examples of what you find frustrating about the UX?

- Navigation with links is completely useless
    - you have to use query syntax in a search bar to do anything.
    - the top bar is always global and resets any parameters
- URL paths are used in non-standard ways (non-encoded slashes appearing as "query params")
    - seems like the resource organization, and by proxy the API, is generally of poor design
- once you add multiple clauses, the header disappears
- navigating to browse code and then review from there removes filters
- paging links are basically MIA

There are more, but this seems sufficient for now.

My opinion is still there is additional complexity and confusion from having the code hosted with two systems and contributions through the multiple systems.

What are the benefits that Gerrit provides which are not available through GitHub?

**Reply by myitcv** [reply]
@verdverm 

> Navigation with links is completely useless

The phrases "Gerrithub sucks", "is completely useless" are unnecessary. Please can we keep the conversation constructive, focussing on specific issues that you have, detailing what you were trying to do at the time etc. 

Have you raised GerritHub issues for any of the problems you describe above, providing repro steps for what you are seeing? 

There has been discussion about having a CUE-specific sub-domain of gerrithub.io, so some tweaks might be possible under that change.



> My opinion is still there is additional complexity and confusion from having the code hosted with two systems and contributions through the multiple systems.

To repeat my point above: anyone who wants to contribute via GitHub is not exposed to the GerritHub review process. Similarly, anyone who wants to contribute via GerritHub does not have to interact with the GitHub process. The minimal cost of maintaining both contribution flows is borne entirely by the maintainers, and given the benefits of Gerrit we are comfortable the benefits currently far outweigh those minimal costs.

> What are the benefits that Gerrit provides which are not available through GitHub?

For example https://twitter.com/_myitcv/status/1369620753029677066 

But to stress once again: we don't need to choose between one flow or the other, because we can (and do) support both.

**Reply by verdverm** [reply]
> Have you raised GerritHub issues for any of the problems you describe above, providing repro steps for what you are seeing?

No, I don't have time to raise issues on every product that I find bugs in and don't use.

> If you have any constructive suggestions on areas of GerritHub that need fixing/could be improved, we can pass those on.

**Reply by verdverm** [reply]
Circling back to the discussion title and

> But to stress once again: we don't need to choose between one flow or the other, because we can (and do) support both.

Which is the primary source and which is mirrored? The original inspiration for the question was to revisit this question. 

GitHub is more widely used and social, so I believe it deserves some consideration. 

How difficult is it swap the sync directions?

**Reply by myitcv** [reply]
Gerrit is the source of truth: GitHub is the mirror. 

> The original inspiration for the question was to revisit this question.

What problem do you think we need to be solving here? What is the impact of that problem?

**Reply by myitcv** [reply]
> How difficult is it swap the sync directions?

A non-trivial amount of tooling would be required to change the sync direction. Hence my question above about what problem you are looking to solve here.

**Reply by verdverm** [reply]
I'm more reflecting on some readings and discussions here. In particular, somewhat mirroring the Slack thread, pushing conversations to GitHub and then asking if it is needed. Also the complexity comment around adding more official containers.

Would it be beneficial to focus the code source and changes around GitHub as well? Can complexity be reduced by removing Gerrit from the picture? Is the added complexity and imperfect sync worth the benefits?

Again, my intention is to have an open discussion around these questions as well as understand the relations and rational to other threads.

**Reply by myitcv** [reply]
> I'm more reflecting on some readings and discussions here. In particular, somewhat mirroring the Slack thread, pushing conversations to GitHub and then asking if it is needed. Also the complexity comment around adding more official containers.

I'm not clear what point you are getting at here - please can you clarify?





> Would it be beneficial to focus the code source and changes around GitHub as well? Can complexity be reduced by removing Gerrit from the picture? Is the added complexity and imperfect sync worth the benefits?

Please see https://github.com/cue-lang/cue/discussions/1197#discussioncomment-1215278. 



> Again, my intention is to have an open discussion around these questions as well as understand the relations and rational to other threads.

I understand, but per https://github.com/cue-lang/cue/discussions/1197#discussioncomment-1226021 I am not clear what additional information/problems you are asking everyone to consider here. Is there a problem that you think we have missed, or misunderstood in terms of its impact?

**Reply by verdverm** [reply]
> I'm not clear what point you are getting at here - please can you clarify?

That like #1160 the intention isn't to solve a problem, it's starting a discussion around removing something because it adds extra overhead or complexity, from the posters point of view.

> Please see #1197 (reply in thread).

Are you referring to the twitter link? If so, not everyone agrees that rebasing is a healthy development practice. It doesn't answer the original question of "What are the benefits that Gerrit provides which are not available through GitHub?" Can you `git push` instead of `git codereview mail`, to sync the gerrit "single commit for a PR" to github?

The dual code hosting is not without drawbacks. (https://github.com/cue-lang/cue/discussions/1197#discussioncomment-1216716 for example). It is certainly more complex than a single code host.

So is the additional complexity worth it?

**Comment by verdverm** [comment]
Is there a way to change GitHub PR workflow so that users contributions are reflected as merged PRs rather than closed w/o merging?

![image](https://user-images.githubusercontent.com/1390600/130327484-4df34ddf-f65e-4a43-807b-1dcbb56463c0.png)

**Reply by myitcv** [reply]
I don't know the answer to this.
